### PR TITLE
Refactor ItemScene::loadFromXml

### PR DESCRIPTION
**Pull request by fschlaepfer**
Refactor the horrible `loadFromXml` function in the `ItemScene` class.

This PR also introduces the helper class `ProgressReporter` which encapsulates information about the current progress and how to report it.

### CODE REVIEW

File: include/helper/progress_helper.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#ifndef PROGRESS_HELPER_H
```
Comment (timoll) [inline include/helper/progress_helper.h]: Why is progress_helper.h in include/helper/ and not in itemframework/include/helper?

File: itemframework/src/item/item_scene.cpp [hunk 2]
```diff
@@ -821,185 +822,250 @@ bool ItemScene::loadFromXml(QDomElement& dom)
...
-                if (reportProgress)  {
-                    emit loadingProgress(progress, "Loading " + type); //show the exact name of what is currently loading
-                }
+        auto const outputs = sourceItem->outputs();
+        auto const inputs  = destinationItem->inputs();
 
-                actchild++;
-                AbstractItem* newItem = PluginManager::instance()->createInstance<AbstractItem>(type);
+        if (sourceOutputIndex < 0 || sourceOutputIndex > outputs.count() - 1) {
+            qCritical() << QString("Couldn't find ouput for connector on item %1. Connector deleted.").arg(sourceItem->name());
```
Comment (timoll) [inline itemframework/src/item/item_scene.cpp:950]: I think we should include something like this somewhere:

http://stackoverflow.com/questions/9255887/stl-function-to-test-whether-a-value-is-within-some-range/9259994#9259994

`if (!interval(0, outputs.count()).contains(sourceOutputIndex) {`

would be more readable.

Comment (fschlaepfer) [reply itemframework/src/item/item_scene.cpp:950]: Good point! It appears we are not using this kind of interval query anywhere else in `ItemScene` so I just added a minimal helper lambda to the `loadFromXml` function.

If it turns out that we could use an actual interval class at multiple places in our code, we should add that class and refactor all occurences in a separate PR.

**Comment by t-moe** [comment]
Since when do we use trailing underscores to mark private members 😄 ?

File: itemframework/src/item/item_scene.cpp [hunk 3]
```diff
@@ -821,185 +822,260 @@ bool ItemScene::loadFromXml(QDomElement& dom)
...
+    });
+
+    QHash<qint32, AbstractItem*> itemIds{};
+
+    ProgressReporter progress{reportingFunction, progressGoal, shouldReportProgress};
+    progress.report();
+
+    // since itemsOut is a pointer, it could also be captured by value, but we
+    // want to express that it is written to.
+    auto extractItem = [&itemIds, &itemsOut,
```
Comment (t-moe) [inline itemframework/src/item/item_scene.cpp]: you should probably swap the lambda names of `extractItem` and `extractGraphicsItem`. 
Or use `extractGraphicsItem` for this one, and `extractAbstractItem` for the other one (currently named `extractGraphicsItem`)

**Comment by t-moe** [comment]
Although this was not introduced in this PR, I feel I should mention it now:

item_scene.cpp:319
```c++
    auto shouldRemove = std::bind(std::logical_or<bool> {},
                                  std::bind(is_temporary_connection_line,
                                            _1),
                                  std::bind(std::logical_not<bool> {},
                                            std::bind(both_ends_in_list,
                                                    _1)));

```

is a lot less readable than

```c++
auto shouldRemove = [&](QGraphicsItem const * item) {
   return (is_temporary_connection_line(item) || ! both_ends_in_list(item));
};
```

You seem to use `std::bind` quite often.

An example where I think it's ok (because it shortens the code quite a bit, while still maintaining good readability):
```c++
auto reportingFunction = std::bind(&ItemScene::loadingProgress, this, _1, _2, QString{""});
```

A (fictional) example where I think  it's overkill to use a `std::bind`:

```c++
std::vector<Item*> container;
std::for_each(container.begin(),container.end(),std::bind(&Item::execute,_1));
```
instead of

```c++
std::vector<Item*> container;
for(auto const& item : container) {
    item->execute();
}
```
(which provides better readability)


I think we all should think twice before using a `std::bind`. Sometimes it makes the code less readable and only a little bit shorter.

As a general rule of thumb I think we should **not** use `std::bind` in the following cases:
* Combined with nested `std::bind` (as in the first example)
* In combination with `std::for_each` and using `std::bind` only to bind the `this` parameter of the member function (as in the last example).

**Comment by fschlaepfer** [comment]
I agree that using `bind` in the case you referenced is indeed less readable than a normal function for most people. However, like you said, that was introduced quite some time ago. The main goal for this PR was to split up the several hundred lines long loop into small and understandable pieces. There are many other things in the code we could (and should) improve but using *too much* abstraction is probably not our biggest problem right now.

I wouldn't say I use `bind` quite often. For the `reportingFunction` it is simply the obvious solution. The only other direct occurence of `bind` in this project is in `project_manager_gui.cpp:86`:

```
GuiManager::instance()->registerCloseHandler(std::bind(&ProjectManagerGui::closeWorkspace,this));
```

which was not written by me.

As for your fictional example, I would simply rewrite

```
std::for_each(container.begin(),container.end(),std::bind(&Item::execute,_1));
```
as
```
std::for_each(container.begin(), container.end(), std::mem_fn(&Item::execute));
```
This, I would argue, is more expressive and to the point than using a range for loop with an explicit `item` variable, which is much more fragile and easier to mess up.

More importantly, I think we should think twice before writing an index-based loop, or a loop with a mutable iterator, etc. Those are not just unreadable but actually *dangerous* and can easily lead to hard to find bugs. Quite often those loops can be expressed much more clearly using standard algorithms which makes them less fragile under change.

**Comment by t-moe** [comment]
+1 for `std::mem_fn`.
There are other occurrences of `std::bind`: item_templates_model.cpp.
But I agree there are not many usages overall.

**Review by t-moe — APPROVED** [review]

**Comment by t-moe** [comment]
Waiting with merge, because of discussion on #23

**Comment by fschlaepfer** [comment]
I said *direct* occurence. The one in `ItemTemplatesModel` you mentioned is wrapped by the template `bind1st`, whose entire point is to encapsulate binding the first argument the a given value, so that you don't have to use `bind` directly. :smile:

**Pull request merged**